src/graph.c

```
@@ -15,91 +15,375 @@
  * along with Butterfly.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-#include <packetgraph/graph.h>
+#include <glib.h>
+#include <packetgraph/packetgraph.h>
 #include "brick-int.h"
+#include "graph-int.h"
 
-static GList *add_brick_to_graph(GList *todo, GList *done,
-				 struct pg_brick *b, struct pg_brick *n,
-				 enum pg_side i, FILE *fd)
-{
-	if (!g_list_find(todo, n) &&
-	    !g_list_find(done, n))
-		todo = g_list_append(todo, n);
-	fprintf(fd, "  \"%s:%s\":%s -- \"%s:%s\":%s\n",
-		b->ops->name,
-		b->name,
-		(i == WEST_SIDE ? "west" : "east"),
-		n->ops->name,
-		n->name,
-		(i == WEST_SIDE ? "east" : "west"));
+static void brick_destroy_cb(gpointer data)
+{
+	pg_brick_destroy(data);
+}
+
+struct pg_graph *pg_graph_new(const char *name, struct pg_brick *explore)
+{
+	struct pg_error *error = NULL;
+	struct pg_graph *ret = g_new(struct pg_graph, 1);
+
+	ret->name = g_strdup(name);
+	ret->all = g_hash_table_new_full(g_str_hash, g_str_equal,
+					 NULL, &brick_destroy_cb);
+	ret->pollable = NULL;
+
+	if (explore) {
+		pg_graph_explore_ptr(ret, explore, &error);
+		if (error)
+			pg_error_free(error);
```

The error should not be silent, so we should return the error instead.